ui/apps/platform/src/Components/CompoundSearchFilter/components/SearchFilterAutocomplete.tsx

```diff
@@ -96,11 +100,18 @@ function SearchFilterAutocomplete({
         []
     );
 
+    const autocompleteSearchString = `${searchTerm}:${filterValue ? `r/${filterValue}` : ''}`;
+    const autocompleteContextString = getRequestQueryStringForSearchFilter(autocompleteContext);
+    const autocompleteQuery =
+        autocompleteContextString !== ''
+            ? [autocompleteContextString, autocompleteSearchString].join('+')
+            : autocompleteSearchString;
```

Comment:
Can we also take into account the currently applied `searchFilter` when constructing the query? I know we are already passing that value to the component above this, and merging the search filter lets us give more accurate results. e.g. If a user has applied `Namespace:stackrox` and then switches to the `Deployment Name` filter, the suggested deployments will only be from the ones in the stackrox namespace.

The current method of doing this is almost a plain merge, the only differences are that we apply the regex search modifiers, and do a little bit of special logic to avoid empty results in specific cases (seen at the top of https://github.com/stackrox/stackrox/blob/master/ui/apps/platform/src/Containers/Vulnerabilities/components/FilterAutocomplete.tsx#L63-L93).

Reply:
I see what you mean. This may complicate things though. Imagine someone does the following:

1. Search by `Deployment Name` and see the autocomplete display `deployment-1` to `deployment-10`
2. They switch to `Namespace Name` and select `namespace-1`
3. They go back to `Deployment Name` and see only `deployment-5` showing up now

I wonder if the user might get confused about what's going on with the autocomplete options. 

I agree it would make things nicer because you'll only see what's relevant, but it might also make things difficult if they wanted to search by `deployment-1` and have to clear filters to see the expected results.

Reply:
Yeah its a trade off for sure, let's see what UX thinks. I believe the downsides of each are:
- Not filtering by the search context will likely result in invalid suggestions that empty the resulting table if chosen
- Filtering by context could lead to unpredictable results, and "disappearing" autocomplete suggestions

Reply:
@dvail @sachaudh Thanks folks for highlighting the pros and cons. This is an interesting discussion. I also mentioned to Brad (a separate concern) that today for Compliance we show auto-complete options for all clusters, as opposed to clusters that are configured under a specific profile. This leads to autocomplete results that are invalid suggestions for the given scope one is viewing. It's a separate issue where we are hitting the general Clusters API list, as opposed to a more contextual per-profile/schedule set of clusters. But I mention it because it seems somewhat related to this experience. 

For context-stacking with filters, that wasn't something I had initially considered, but we have examples of that in views like Dashboards and Network Graph where a parent filter influences the available children. Would the same be applicable if I start with a deployment filter and then search for a namespace? Will the namespace autocompletes be restricted by the initial deployment query?

And if I had to think of a third approach, it could maybe be a combination of the two, where we prioritize the contextually relevant options in a group entitled "SOMETHING", and separate "all other matching results" in another group below, so we don't hide autocomplete suggestions, but also encourage the more sensible set.

I don't know if there are edge cases with this frankly, and if it opens a can of worms too. But if I had to lean one way given what I surfaced for the Compliance section, is that ideally we don't surface autocomplete results for invalid objects that are not going to be found on the table, as it does feel a bit broken when we do that.

Reply:
I tried it out and yes the namespace autocomplete would be restricted by the initial deployment query.

I like your suggestion about grouping by relevant and non-relevant autocomplete options. I think that would be helpful. I'm worried about the scope of this change now and how we will need proper testing to make sure things work. I'll try it out and see how it goes, but I will just call out that it might be a bit more extra work and will take a little bit of time to also think about edge cases.

Reply:
@sachaudh thanks Saif, and so for that reason, if we want to start with Dave's suggestion for now, and maybe test the third option after 4.5, with more room to breathe, that is okay too. I defer to your judgment.

Reply:
ah shoot, one more question @sachaudh @dvail  - If I filter for a deployment, and then another deployment -- I am guessing those are ORs since they are values contained to the same key and the former will not affect/filter out the autocomplete of the latter. Is that accurate?

Reply:
@alanonthegit Regarding the last question, yes this is how this will be intended to work and how it works with the Workload CVE filters pre-advanced filters (just tested on the dogfooding instance to be sure).

IMO I think the context stacking is something we definitely want for 4.5, as otherwise I think autocomplete in general will be of limited use. The separate groupings for autocomplete results though should be investigated post 4.5 as that will be a relatively large change to how filters are currently made.

Reply:
Thanks @dvail @sachaudh  agreed, I think the trade-offs of not showing all the results are less damaging than showing results that don't match what is available in the table set. It would appear as though the feature is broken. 

A 4th approach, which doesn't require displaying both sets of data, might be to include a tiny snippet of text at the bottom of the autocomplete that explains "Available options are determined by prior filters. Clear all filters to search the entire dataset". Or something shorter to that extent. But I think this might be overkill.

Reply:
I'll try out the context stacking for now. We can discuss the visual adjustments and enhancements more during 4.6

Reply:
I was thinking about this more and was wondering if the following would be an issue:

Context: **NamespaceA** has **DeploymentA**, **DeploymentB**, and **DeploymentC**. **NamespaceB** has **DeploymentD** and **DeploymentE**.

1. The user navigates to a page displaying a list of deployments.
2. The user uses advanced filters to search by **NamespaceA**, so the table now only shows **DeploymentA**, **DeploymentB**, and **DeploymentC**.
3. When the user goes to the **Deployment Name** filter, the autocomplete only shows **DeploymentA**, **DeploymentB**, and **DeploymentC**.
4. The user wants to search for **DeploymentD**.
5. This becomes an issue with context filtering for autocomplete options. Ideally, the user wants to see deployments for **NamespaceA** and also **DeploymentD** from **NamespaceB**, but the autocomplete now limits their ability to do that.

Reply:
@sachaudh @dvail the current UI behavior when I get to step 4, is that I show `Deployment D` in the dropdown. However, filtering for  `Deployment D` while looking at an already pre-filtered table of `Namespace A` returns no results. Since these 2 values are being ANDed, then I don't think it's as much of a concern?

We basically keep the auto-complete options in parity with the data we display in the table, so we avoid presenting autocomplete entries that will return zero results.

<img width="1279" alt="image" src="https://github.com/stackrox/stackrox/assets/28272783/bdefcede-3b15-445b-997f-3e7ec5e42ccd">

<img width="1385" alt="image" src="https://github.com/stackrox/stackrox/assets/28272783/3a71ca8e-bc83-4dcb-963e-61d74f43123a">

Reply:
+1 - I think the potential confusion of items missing from the autocomplete is necessary in order to present valid options to the user when filtering across categories. The alternative of showing everything means that once a user applies a filter in one category, there is a strong likelihood that all autocomplete values for a different category will become invalid. This likelihood increases with the size of the environment. i.e. In an environment with 50,000 deployments, a user who filters first by Namespace and then Deployments will almost certainly get a list of 10 suggested deployments where each choice would cause the table to be empty.

Reply:
@alanonthegit if you switch to the `images without vulnerabilities` section, it'll show up. It should OR the values. What you're seeing is exactly the issue that led to this PR 👀

Reply:
But yeah I guess the situation I raised might be ok. If the user goes through filtering from a top-down approach, they'll add additional filters as they scope down the table. If the user goes through filtering from a more random approach, then I guess it wouldn't be as helpful for them.